pygfx/helpers/_gizmo.py

```diff
@@ -0,0 +1,747 @@
+    * Click the center sphere to toggle between object-space, world-space and screen-space.
+    * Grab the center sphere for uniform scaling.
+    * Grab the cubes for one-directional scaling (only in object-space).
+    * Grab the arrows to translate in one direction.
+    * Grab the planes to translate in two directions.
+    * Grab the spheres to rotate.
+    """
+
+    def __init__(self, object=None, screen_size=100):
+        super().__init__()
+
+        # We store these as soon as we get a call in handle_event
+        self._renderer = None
+        self._canvas = None
+        self._camera = None
+
+        # A dict that stores the state at the start of a drag. Or None.
+        self._ref = None
+
+        # The (approximate) size of the gizmo on screen
+        self._screen_size = float(screen_size)
+
+        # Init
+        self.set_object(object)
+        self._create_elements()
+        self.toggle_mode("object")  # object, world, screen
+        self._highlight()
+
+    def set_object(self, object):
+        """Set the WorldObject to control with the gizmo."""
+        if object is None:
+            self._object_to_control = None
+        else:
+            assert isinstance(object, WorldObject)
+            self._object_to_control = object
+
+    def toggle_mode(self, mode=None):
+        """Toggle between modes. If the mode is omitted, will move to the next mode.
+        Must be one of "object", "world", or "screen".
+        """
+        modes = "object", "world", "screen"
+        if not mode:
+            mode = {"object": "world", "world": "screen"}.get(self._mode, "object")
+        if mode not in modes:
+            raise ValueError(f"Invalid mode '{mode}', must be one of {modes}.")
+        self._mode = mode
+        self._on_mode_switch()  # The gizmo looks a bit different in each mode
+
+    def _create_elements(self):
+        """Create lines, handles, etc."""
+
+        # The location of the scale handles, and where the arcs meet the lines.
+        self._halfway = halfway = 0.65
+
+        # --- the parts that are at the center
+
+        # Create screen translate handle
+        sphere_geo = sphere_geometry(0.07)
+        scale_uniform = Mesh(
+            sphere_geo,
+            MeshBasicMaterial(color="#ffffff"),
+        )
+        scale_uniform.scale.set(1.5, 1.5, 1.5)
+
+        # --- the parts that are fully in one dimension
+
+        # Create lines
+        line_geo = Geometry(positions=[(0, 0, 0), (halfway, 0, 0)])
+        line_x = Line(
+            line_geo,
+            LineMaterial(thickness=4, color=RED),
+        )
+        line_y = Line(
+            line_geo,
+            LineMaterial(thickness=4, color=GREEN),
+        )
+        line_z = Line(
+            line_geo,
+            LineMaterial(thickness=4, color=BLUE),
+        )
+
+        # Create 1D translate handles
+        cone_geo = cone_geometry(0.1, 0.17)
+        cone_geo.positions.data[:] = cone_geo.positions.data[:, ::-1]
```

Comment:
what does this line do?

Reply:
Reverses xyz -> zyx, to put the cones in the correct orientation. Added a comment.